### PR TITLE
Implement `s.example()` in terms of `@given`, not `find()`

### PR DESCRIPTION
**Pull request by Zac-HD**
This makes the outputs from `.example()` more representative of what will happen in tests - which I've found is a common point of confusion in workshops - and makes it easier to finally deprecate `find()` (#1694) in my next PR!

### CODE REVIEW

File: hypothesis-python/src/hypothesis/searchstrategy/strategies.py [hunk 1]
```diff
@@ -284,37 +290,25 @@ def example(self, random=None):
...
+                phases=(Phase.generate,),
+                suppress_health_check=HealthCheck.all(),
             )
-        except (NoSuchExample, Unsatisfiable):
-            # This can happen when a strategy has only one example. e.g.
-            # st.just(x). In that case we wanted the first example after all.
-            if first:
-                return first[0]
-            raise NoExamples(u"Could not find any valid examples in 100 tries")
+            def engine_inside_example(ex):
```
Comment (DRMacIver) [inline hypothesis-python/src/hypothesis/searchstrategy/strategies.py]: nit: This is a weird name for it.

File: hypothesis-python/src/hypothesis/searchstrategy/strategies.py [hunk 2]
```diff
@@ -284,37 +290,25 @@ def example(self, random=None):
...
+                self.__examples.append(ex)
+
+            self.__examples = []
+            engine_inside_example()
+            self.__examples.reverse()
+        return self.__examples.pop()
```
Comment (DRMacIver) [inline hypothesis-python/src/hypothesis/searchstrategy/strategies.py]: I think it might be worth shuffling the examples at the point of generation. This means you get a better representation of what given actually does, and it guards against the case where you only try it once (in which case it always gives you the minimal example) either deliberately or because you're trying a strategy that is recreated each time.

In fact, I'm not sure I'm keen on keeping state around on the strategy objects to support what is a very niche use case whose performance mostly shouldn't matter. `10` isn't a very large number, so why not just generate all `10` each time and return a randomly sampled one?

File: hypothesis-python/tests/cover/test_core.py [hunk 3]
```diff
@@ -23,21 +23,20 @@
 import hypothesis.strategies as s
 from hypothesis import find, given, reject, settings
 from hypothesis.errors import NoSuchExample, Unsatisfiable
+from tests.common.debug import minimal
```
Comment (DRMacIver) [inline hypothesis-python/tests/cover/test_core.py]: I think all the changes in this file are from an unrelated PR?

File: hypothesis-python/tests/nocover/test_find.py [hunk 4]
```diff
@@ -21,8 +21,8 @@
 
 import pytest
 
-from hypothesis import find, settings as Settings
-from hypothesis.errors import NoSuchExample
```
Comment (DRMacIver) [inline hypothesis-python/tests/nocover/test_find.py]: I think these are also unrelated changes.

File: hypothesis-python/tests/cover/test_core.py (hunk 3 above)
Comment (Zac-HD) [reply hypothesis-python/tests/cover/test_core.py]: Whoops, yeah - they belong in the deprecation-of-`find` PR, not this one.

File: hypothesis-python/src/hypothesis/searchstrategy/strategies.py (hunk 1 above)
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/searchstrategy/strategies.py]: I've made it more descriptive; and added a comment - it's long because this name can appear in user-facing tracebacks, so I prefer to make it somewhat descriptive.
